integration_test/src/test_runner/mod.rs

```diff
@@ -0,0 +1,327 @@
+}
+
+#[derive(Debug, Error, PartialEq, Eq)]
+pub enum Error {
+    /*
+    #[error("test runner failed for test {test}: expected {expected:?}, actual {actual:?}")]
+    IntegrationTest {
+        test: String,
+        expected: String,
+        actual: String,
+    },
+     */
+    #[error("unable to read file to string: {0}")]
+    CannotReadFileToString(String),
+    #[error("failed to read file: {0}")]
+    InvalidFile(String),
+    #[error("failed to read directory: {0}")]
+    InvalidDirectory(String),
+    #[error("failed to load file paths: {0}")]
+    InvalidFilePath(String),
+    #[error("unable to deserialize YAML file: {0}")]
+    CannotDeserializeYaml(String),
+    #[error("missing query or function in test: {0}")]
+    MissingQueryOrFunction(String),
+    #[error("unsuccessful SQL return code encountered: {0}")]
+    SqlReturn(String),
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct IntegrationTest {
+    pub tests: Vec<TestEntry>,
+}
+
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
+pub struct TestEntry {
+    pub description: String,
+    pub db: String,
+    pub query: Option<String>,
+    pub function: Option<Vec<Value>>,
+    pub expected_result: Option<Vec<BTreeMap<String, Value>>>,
+    pub skip_reason: Option<String>,
+    pub ordered: Option<bool>,
+    pub expected_column_name: Option<Vec<String>>,
+    pub expected_sql_type: Option<Vec<String>>,
+    pub expected_bson_type: Option<Vec<String>>,
+    pub expected_precision: Option<Vec<i32>>,
+    pub expected_scale: Option<Vec<i32>>,
+    pub expected_nullability: Option<Vec<String>>,
+}
+
+/// load_file_paths reads the given directory and returns a list its file path
+/// names.
+pub fn load_file_paths(dir: PathBuf) -> Result<Vec<String>, Error> {
+    let mut paths: Vec<String> = vec![];
+    let entries = fs::read_dir(dir).map_err(|e| Error::InvalidDirectory(format!("{:?}", e)))?;
+    for entry in entries {
+        match entry {
+            Ok(de) => {
+                let path = de.path();
+                if (path.extension().unwrap() == "yml") || (path.extension().unwrap() == "yaml") {
+                    paths.push(path.to_str().unwrap().to_string());
+                }
+            }
+            Err(e) => return Err(Error::InvalidFilePath(format!("{:?}", e))),
+        };
+    }
+    Ok(paths)
+}
+
+/// parse_test_file_yaml deserializes the given YAML file into a
+/// IntegrationTest struct.
+pub fn parse_test_file_yaml(path: &str) -> Result<IntegrationTest, Error> {
+    let mut f = fs::File::open(path).map_err(|e| Error::InvalidFile(format!("{:?}", e)))?;
+    let mut contents = String::new();
+    f.read_to_string(&mut contents)
+        .map_err(|e| Error::CannotReadFileToString(format!("{:?}", e)))?;
+    let yaml: IntegrationTest = serde_yaml::from_str(&contents)
+        .map_err(|e| Error::CannotDeserializeYaml(format!("{:?}", e)))?;
+    Ok(yaml)
```

Comment:
```suggestion
    let f = fs::File::open(path).map_err(|e| Error::InvalidFile(format!("{:?}", e)))?;
    let integration_test: IntegrationTest = serde_yaml::from_reader(f).map_err(|e| Error::CannotDeserializeYaml(...))?;
    Ok(integration_test)
```
I've been using `serder_yaml::from_reader` directly on the `File` instead of going to `String` as an intermediate step. Is there a reason you have the `File` -> `String` step before going from `String` -> data structure?

Reply:
I tried the suggested method and it worked as expected.  Good to know the intermediate `String` step is not needed.
I've made this change 👍